lib/trino-hive-formats/src/test/java/io/trino/hive/formats/avro/TestAvroPageDataReaderWithAvroNativeTypeManagement.java

```diff
@@ -265,37 +275,116 @@ public void testWriting()
         }
     }
 
+    @Test
+    public void testNegativeEpochTimestampsAndDates()
+            throws IOException, AvroTypeException
+    {
+        Location testLocation = createLocalTempLocation();
+        Schema timestampsSchema = SchemaBuilder.builder()
+                .record("timestamps")
+                .fields()
+                .name("timestampMillis")
+                .type(TIMESTAMP_MILLIS_SCHEMA).noDefault()
+                .name("timestampMicros")
+                .type(TIMESTAMP_MICROS_SCHEMA).noDefault()
+                .name("timestampNanos")
+                .type(TIMESTAMP_NANOS_SCHEMA).noDefault()
+                .name("date")
+                .type(DATE_SCHEMA).noDefault()
+                .endRecord();
+        ImmutableList.Builder<Block> blocks = ImmutableList.builder();
+
+        long nanosBeforeEpoch = -1_000_000_000L;
+        SqlTimestamp writeTimestamp = SqlTimestamp.newInstance(9, nanosBeforeEpoch / Timestamps.NANOSECONDS_PER_MICROSECOND, 0);
+
+        BlockBuilder timestampMillisBlock = TimestampType.TIMESTAMP_MILLIS.createFixedSizeBlockBuilder(1);
+        TimestampType.TIMESTAMP_MILLIS.writeLong(timestampMillisBlock, nanosBeforeEpoch / Timestamps.NANOSECONDS_PER_MICROSECOND);
+        blocks.add(timestampMillisBlock.build());
+
+        BlockBuilder timestampMicrosBlock = TimestampType.TIMESTAMP_MICROS.createFixedSizeBlockBuilder(1);
+        TimestampType.TIMESTAMP_MICROS.writeLong(timestampMicrosBlock, nanosBeforeEpoch / Timestamps.NANOSECONDS_PER_MICROSECOND);
+        blocks.add(timestampMicrosBlock.build());
+
+        BlockBuilder timestampNanosBlock = TimestampType.TIMESTAMP_NANOS.createFixedSizeBlockBuilder(1);
+        TimestampType.TIMESTAMP_NANOS.writeObject(timestampNanosBlock, new LongTimestamp(writeTimestamp.getEpochMicros(), writeTimestamp.getPicosOfMicros()));
+        blocks.add(timestampNanosBlock.build());
+
+        int daysBeforeEpoch = -3;
+        BlockBuilder dateBlock = DateType.DATE.createFixedSizeBlockBuilder(1);
+        DateType.DATE.writeInt(dateBlock, daysBeforeEpoch);
+        blocks.add(dateBlock.build());
+
+        Page timestampsPage = new Page(blocks.build().toArray(Block[]::new));
+        try (AvroFileWriter fileWriter = new AvroFileWriter(
+                trinoLocalFilesystem.newOutputFile(testLocation).create(),
+                timestampsSchema,
+                new NativeLogicalTypesAvroTypeManager(),
+                AvroCompressionKind.NULL,
+                ImmutableMap.of(),
+                timestampsSchema.getFields().stream().map(Schema.Field::name).collect(toImmutableList()),
+                new NativeLogicalTypesAvroTypeBlockHandler().typeFor(timestampsSchema).getTypeParameters(), false)) {
+            fileWriter.write(timestampsPage);
+        }
```

Comment:
This can't be the only test. `NativeLogicalTypesAvroTypeManager` is our class and its logic may or may not be entirely correct.

There should be a test with hive writing the file. See example existing test https://github.com/trinodb/trino/blob/16b41f68e40b92fefce08fdc2b21b238f5e7f48b/testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestAvroSchemaLiteral.java#L52-L53